app/bundles/CampaignBundle/Tests/Functional/Campaign/ContactCampaignHistoryOrderTest.php

```diff
@@ -0,0 +1,164 @@
+        $timelineEvents = $engagements['events'];
+
+        $historyOrder = [];
+        foreach ($timelineEvents as $timelineEvent) {
+            $historyOrder[] = trim(explode('/', $timelineEvent['eventLabel']['label'])[0]);
+        }
+
+        Assert::assertSame(
+            $expectedOrder,
+            $historyOrder,
+            'The campaign history is not sorted by creation order in '.$orderByDir
+        );
+    }
+
+    /**
+     * @return iterable<string, array{0: string, 1: string[]}>
+     */
+    public static function provideTimelineOrderData(): iterable
+    {
+        $expectedOrder = [
+            'Update Contact',
+            'Adjust Points',
+            'Add to Company',
+        ];
+
+        yield 'ASC' => [
+            'ASC',
+            $expectedOrder,
+        ];
+
+        yield 'DESC' => [
+            'DESC',
+            array_reverse($expectedOrder),
+        ];
+    }
+
+    private function createCampaignWithEvents(LeadList $segment): Campaign
+    {
+        // Dummy company for Add to Company action
+        $company = new Company();
+        $company->setName('Test Company');
+        $this->em->persist($company);
+        $this->em->flush();
+
+        $campaign = new Campaign();
+        $campaign->setName('History Order Test');
+        $campaign->setIsPublished(true);
+        $campaign->setPublishUp(new \DateTime('-1 day'));
+        $campaign->addList($segment);
+
+        // Action 1: Update Contact
+        $event1 = new Event();
+        $event1->setName('Update Contact');
+        $event1->setType('lead.updatelead');
+        $event1->setProperties(['tags' => ['test-tag']]);
+        $event1->setEventType(Event::TYPE_ACTION);
+        $event1->setCampaign($campaign);
+        $campaign->addEvent(1, $event1);
+
+        // Action 2: Adjust Points
+        $event2 = new Event();
+        $event2->setName('Adjust Points');
+        $event2->setType('lead.changepoints');
+        $event2->setProperties(['points' => 10]);
+        $event2->setEventType(Event::TYPE_ACTION);
+        $event2->setCampaign($campaign);
+        $event2->setParent($event1);
+        $campaign->addEvent(2, $event2);
+
+        // Action 3: Add to Company
+        $event3 = new Event();
+        $event3->setName('Add to Company');
+        $event3->setType('lead.addtocompany');
+        $event3->setProperties(['company' => $company->getId()]);
+        $event3->setEventType(Event::TYPE_ACTION);
+        $event3->setCampaign($campaign);
+        $event3->setParent($event2);
+        $campaign->addEvent(3, $event3);
+
+        $this->em->persist($campaign);
+
+        return $campaign;
+    }
+
+    private function createContactAndAddToSegment(LeadList $segment): Lead
+    {
+        $contact = new Lead();
+        $contact->setEmail('history-test@mautic.com');
+        $this->em->persist($contact);
+        $this->em->flush();
+
+        /** @var \Mautic\LeadBundle\Model\ListModel $listModel */
```

Comment:
```suggestion
        /** @var ListModel $listModel */
```
plus
```php
use Mautic\LeadBundle\Model\ListModel;
```
nitpick: Minor thing